lib/backend/syncersv1/updateprocessors/processor.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package updateprocessors
+
+import (
+	"fmt"
+
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/backend/watchersyncer"
+)
+
+// NewGeneralUpdateProcessor implements an update processor that only needs to take in
+// a conversion function. This is only meant to be used in cases where the Process and
+// OnSyncerStarting methods do not have any special handling.
+//
+// One of the main differences between the general update processor and the conflict
+// resolving name cache processor is that the conflict resolving name cache processor
+// handles deletes by inspecting the value in the key-value pair. The general update processor
+// will strictly convert the information it is given, whether the value is nil or not.
+
+func NewGeneralUpdateProcessor(v2Kind string, converter ConvertV2ToV1) watchersyncer.SyncerUpdateProcessor {
+	return &generalUpdateProcessor{
+		v2Kind:    v2Kind,
+		converter: converter,
+	}
+}
+
+type generalUpdateProcessor struct {
+	v2Kind    string
+	converter ConvertV2ToV1
+}
+
+func (gup *generalUpdateProcessor) Process(kvp *model.KVPair) ([]*model.KVPair, error) {
```

Comment:
As it stands, I'm not sure this generalUpdateProcess buys us much.  The only common processing is the validation of the resource Kind and then the rest of the processing is done in the converter.  And from what I can see, you do the type casting an Kind validation again in each of the individual converter functions.

It might be more useful if it took two converter functions, one to handle the key, and one to handle the value.  That way the handling of a delete can be handled generically here by only calling the convertKey method - whereas for an add/update you call both.

Also, since you are already asserting that the v2 key is a ResourceKey, the function signature for the convert key method could look more like:

`type ConvertV2KeyToV1 func(v2Key model.ResourceKey) model.Key`

and the value converter I guess would look like:

`type ConvertV2ResToV1 func(v2Obj runtime.Object) (interface{}, error)`

Reply:
Also, perhaps rather than GeneralUpdateProc... maybe SimpleUpdateProc... would be better since it can only be used for simple cases where the keys are directly mappable from v2 to v1.